Add a high-level overview of MyST components (#1950)
jupyter-book/mystmd: docs/about/overview.md

# What is the MyST stack?

This section provides a high-level overview of the main concepts in the MyST ecosystem, as well as how it relates to specific tools.

MyST might refer to several things, depending on the context.
Here are a few key concepts in MyST:

**The MyST Specification**. A specification that describes the structure and function of MyST documents. It provides a framework for defining all of the kinds of content and metadata that a MyST document can contain. For example, the MyST specification defines how sections can contain paragraphs, how paragraphs can contain sentences, and how sentences can contain a "bolded" chunk of text. The specification is [defined at mystmd.org/spec](https://mystmd.org/spec).

> **A MyST Document**. A document that adheres to the MyST Specification. It is usually in the form of an [Abstract Syntax Tree](https://en.wikipedia.org/wiki/Abstract_syntax_tree) that is structured in JSON (or some other machine-readable form), where every bit of content has been parsed and tagged according to the MyST Specification.

Reviewer: The MyST AST is a supserset of MDAST, in theory (although formally there are a few bits that need fixing).

> # What is the MyST stack?

Reviewer 2: Also, importantly, talks about the relationship between MyST and Jupyter Book. Worth mentioning in the title or first paragraph?

Reviewer: We have a little of that content [here](https://mystmd.org/guide/background#how-do-jupyter-book-and-mystmd-relate) too, so we might be fine not to address it in this title.

This section provides a high-level overview of the main concepts and tools in the MyST ecosystem. See the [guiding principles of the MyST ecosystem](../guiding-principles.md) for some more high-level context.

Here are a few key concepts in MyST:

(myst-markdown)=

**MyST Markdown**. MyST stands for "Markedly Structured Text". It is a flavor of [markdown](https://en.wikipedia.org/wiki/Markdown) that was designed to make it easy to write content that can be parsed by a [MyST Document Engine](#myst-engine). MyST Markdown syntax was created to support functionality in the MyST specification like extensions, roles and directives, and cross-references. It was originally created for Jupyter Book via the [MyST Parser for Sphinx](https://myst-parser.readthedocs.io), but now serves as a standard markdown syntax across the MyST ecosystem.

(myst-engine)=

> **A MyST Document Engine**. A MyST Document Engine parses text-based documents into [MyST AST](#myst-ast) according to the [MyST Specification](#myst-specification)[^myst]. For example, a MyST Document Engine might know how to transform a text file with [MyST Markdown](#myst-markdown) into a `.json` output that follows the [MyST Specification](#myst-spec). The official MyST Document Engine is documented at [mystmd.org/guide](https://mystmd.org/guide), and is maintained by the [Jupyter Book team](https://compass.jupyterbook.org)

Reviewer: This is very pedantic, but I think it's useful to define the MyST engine as something that builds AST. As we grow the number of supported input formats, the notion of what the MyST engine consumes is less and less relevant.

```suggestion
**A MyST Document Engine**. A MyST Document Engine builds [MyST AST](#myst-ast) according to the [MyST Specification](#myst-specification)[^myst]. For example, a MyST Document Engine might know how to transform a text file written in [MyST Markdown](#myst-markdown) into JSON output that follows the [MyST Specification](#myst-spec). The official MyST Document Engine is documented at [mystmd.org/guide](https://mystmd.org/guide), and is maintained by the [Jupyter Book team](https://compass.jupyterbook.org)
```

> [^myst]: Most commonly, MyST engines parse [MyST Markdown](#myst-markdown). However, a MyST engine can parse any kind of text syntax as long as the _result_ is a [MyST AST](#myst-ast) that follows the [MyST specification](#myst-spec). This is why the [official MyST Document Engine](https://mystmd.org) can parse other kinds of markdown syntax, like some of Pandoc's syntax.

Reviewer: Simpler to refer to LaTeX here as an alternative input format.
```suggestion
[^myst]: Often we focus upon a MyST Engine's ability to parse [MyST Markdown](#myst-markdown). However, a MyST Engine may consume _any_ kind of input markup as long as the _result_ is a [MyST AST](#myst-ast) that follows the [MyST specification](#myst-spec). This is why the [official MyST Document Engine](https://mystmd.org) can parse other kinds of markup, such as a subset of LaTeX.
```

(myst-specification)=

> **The MyST Specification**. A specification that describes the structure and function of MyST documents. It provides a framework for defining all of the kinds of content and metadata that a MyST document can contain. For example, the MyST specification defines how sections can contain paragraphs, how paragraphs can contain sentences, and how sentences can contain a "bolded" chunk of text. The specification is [defined at mystmd.org/spec](https://mystmd.org/spec).

Reviewer 2: Does this describe MyST Markdown, as suggested in the last paragraph, or does it talk more about the structure for creating, e.g., a notebook document around MyST Markdown content?

If the first, this could perhaps be simplified a bit along the lines of "this is the official reference definition of the MyST Markdown syntax".

Reviewer: The MyST spec is nominally two things: the AST, and the markup grammar. However, increasingly I _think_ we're moving to the AST being the spec, and the grammar being recognised independently. Would welcome the rest of @jupyter-book/core-team's view there. For me, the benefits are that we can tighten the concept of the AST being the "unit of exchange and manipulation", and avoid tying ourselves to a language that is ultimately just another markup language.[^nb]

[^nb]: In the sense that ultimately MyST could operate with something like djot, rst, qmd, or even LaTeX without adversely modifying the engine underneath.

> **The MyST Specification**. A specification that describes the structure and function of MyST documents. It provides a framework for defining all of the kinds of content and metadata that a MyST document can contain. For example, the MyST specification defines how sections can contain paragraphs, how paragraphs can contain sentences, and how sentences can contain a "bolded" chunk of text. The specification is [defined at mystmd.org/spec](https://mystmd.org/spec).

Reviewer: Use an xref?
```suggestion
**The MyST Specification**. A specification that describes the structure and function of MyST documents. It provides a framework for defining all of the kinds of content and metadata that a MyST document can contain. For example, the MyST specification defines how sections can contain paragraphs, how paragraphs can contain sentences, and how sentences can contain a "bolded" chunk of text. The specification is [defined at mystmd.org/spec](xref:spec#overview).
```

(myst-ast)=

> **MyST Abstract Syntax Tree (AST)**. A bundle of parsed content that adheres to the MyST Specification. An [Abstract Syntax Tree](https://en.wikipedia.org/wiki/Abstract_syntax_tree) is a way of structuring information according to a formal specification, so MyST Document Engines know how to parse content and _create_ an AST out of that content. Usually, a MyST AST takes the form of a `.json` data structure, with metadata attached to each piece of content that describes its role, relationships with other content, etc. Because MyST AST follows a formal structure, it can be **rendered** into many kinds of outputs (like HTML, PDF, docx, etc).

Reviewer 2: a way of structuring -> a way of representing

An AST represents information, I suppose, but it feels like there may be a bit more precise phrasing for that. Something like:

"An AST is a standardized (formal?), structured representation of content and markup."

Why does it exist? "... for internal use by computer programs" — some notion of "it's not user facing, but we use it to pass standardized representations between computer programs.

And the ultimate motivation? "The AST allows us to de-couple the steps of parsing markdown and generating output." or, perhaps clearer, "The AST allows us to separate the steps of parsing markdown, and of then generating output from the parsed content."

> **MyST Abstract Syntax Tree (AST)**. A bundle of parsed content that adheres to the MyST Specification. An [Abstract Syntax Tree](https://en.wikipedia.org/wiki/Abstract_syntax_tree) is a way of structuring information according to a formal specification, so MyST Document Engines know how to parse content and _create_ an AST out of that content. Usually, a MyST AST takes the form of a `.json` data structure, with metadata attached to each piece of content that describes its role, relationships with other content, etc. Because MyST AST follows a formal structure, it can be **rendered** into many kinds of outputs (like HTML, PDF, docx, etc).

Reviewer: ```suggestion
**MyST Abstract Syntax Tree (AST)**. A structured representation of a MyST Document, typically built from a markup language (like MyST Markdown) by a MyST Document Engine, that adheres to the MyST Specification. Usually, a MyST [Abstract Syntax Tree (AST)](wiki:Abstract_syntax_tree) is represented as a JSON data structure, with metadata attached to each piece of content that describes its role, relationships with other content, etc. Because MyST AST follows a formal structure, it can be **rendered** into many kinds of outputs (like HTML, PDF, docx, etc).
```

Here's a diagram showing how these all relate to one another:

> :::{figure} ../images/myst-diagram.svg

Reviewer 2: I like SVG diagrams and draw most diagrams in Inkscape myself. That said, MyST does support Mermaid, so that's another option that makes the diagram easier to update.

Reviewer: For now, let's leave it as a static figure — I think we have a few more of these, and it would be sensible to do a standardisation pass later on.

Author: Just noting that this figure is made with Excalidraw, and has all of the content inside it needed to edit the figure again. If you upload that SVG into the Excalidraw app (or are running the app locally) then you'll be able to edit anything in the figure and then update the SVG with a new PR.

Would it help to document that process? I find it super useful!

An overview of some major parts of the MyST stack and how they relate to one another.

> In the most common workflow, a text file written in MyST Markdown is parsed by the [MyST Document Engine](https://mystmd.org/guide) which outputs MyST Document that follows the [MyST Specification](https://mystmd.org/spec) (usually a JSON file). This MyST Document can be rendered into many different kinds of outputs, most-commonly HTML or PDF. This usually done by the MyST Document Engine, but can be done by any application that understands how to parse and use the MyST Document Specification.

Reviewer 2: typo: outputs MyST Document. What does it output, MyST AST?

> In the most common workflow, a text file written in MyST Markdown is parsed by the [MyST Document Engine](https://mystmd.org/guide) which outputs MyST Document that follows the [MyST Specification](https://mystmd.org/spec) (usually a JSON file). This MyST Document can be rendered into many different kinds of outputs, most-commonly HTML or PDF. This usually done by the MyST Document Engine, but can be done by any application that understands how to parse and use the MyST Document Specification.

Reviewer: I agree that the word document is probably too overloaded in the context of MyST-MD.
```suggestion
In the most common workflow, a text file written in MyST Markdown is parsed by the [MyST Document Engine](https://mystmd.org/guide) which outputs MyST AST that follows the [MyST Specification](https://mystmd.org/spec) (usually a JSON file). This MyST AST can be rendered into many different kinds of outputs, most-commonly HTML or PDF. This usually done by the MyST Document Engine, but can be done by any application that understands how to parse and use the MyST Document Specification.
```

:::

## Are there other MyST engines?

The [MyST Specification](https://mystmd.org/spec) and [MyST Markdown Syntax](https://mystmd.org/guide) was designed so that others could implement their own parsers, document engines, and renderers using MyST. The best example of another MyST Document Engine is the MyST parser for Sphinx.

The [MyST Parser for Sphinx](https://myst-parser.readthedocs.io) is an extension for [Sphinx](https://sphinx-doc.org) that can parse MyST Markdown into Sphinx's internal document structure. It was created to allow users to parse MyST Markdown syntax into Sphinx for [V1 of Jupyter Book](https://jupyterbook.org). It now exists as an independent extension for the Sphinx community, as Jupyter Book now uses the MyST Document Engine.

The MyST Parser for Sphinx does not support all of the functionality that is defined in the MyST Specification - that's OK! Its goal is to be a useful tool for the Sphinx community that only leverages parts of the MyST ecosystem (the markdown flavor).

## How does Jupyter Book relate to MyST?

> The MyST Markdown Syntax was originally created to allow [Jupyter Book V1](https://jupyterbook.org) to use the [Sphinx Document engine](https://sphinx-doc.org) with markdown content.

Reviewer 2: This seems to say: "We created a new version of Markdown so we could use Markdown with Sphinx".

The actual motivation may be a bit tricky to explain, so also fine to leave as-is :)

Author: that is kind of true basically. Basically:

- Jupyter Book wanted to use Sphinx for its engine
- But Sphinx only spoke rST and we didn't want that
- And roles/directives didn't exist as a concept in a pre-existing markdown flavor
- So we created MyST as a Sphinx-native markdown flavor.

Over time, the Jupyter Book team decided that the most sustainable path forward was to create its own document engine. This became [MyST Document Engine](https://mystmd.org/guide)[^cn]. It serves as the engine behind Jupyter Book V2.

You can think of Jupyter Book as a _distribution of the MyST Document Engine_. In other words, Jupyter Book wraps the MyST Document Engine application, with out-of-the-box configuration that supports a multi-page community knowledge base or documentation site. MyST is created and maintained by the [Jupyter Book team](https://compass.jupyterbook.org).

> Currently, Jupyter Book and the MyST Document Engine share much of the same functionality. Over time, we imagine that the MyST Document Engine will be a more flexible tool that makes fewer assumptions about the end use-case that it's being used for. Jupyter Book will then have "more opinions" above the base configuration of the MyST Document Engine. That said - we are still figuring this out! We aren't sure the exact relationship the two will have, and will revisit this as we learn more.

Reviewer 2: This is great. I think the more standard phrasing is "Jupyter Book will be a more opinionated ...". Dash after "said" should be em-dash, or could be replaced by a comma.

Reviewer: I'm so here for someone who pushes em-dashes!

> [^cn]: The official MyST Document Engine was actually originally created by [Curvenote](https://curvenote.com), which donated their internal document engine to the Jupyter Book project and became co-maintainers of the Jupyter Book stack moving forward.

Reviewer 2: I would inline this. It's an important part of the heritage.
